M1 installation instructions using the x86 build of Lean (#231)
leanprover-community/leanprover-community.github.io: templates/install/macos.md

1. Open a new terminal window and install XCode Command Line Tools and Rosetta 2 using `xcode-select --install` and `softwareupdate --install-rosetta`.

2. We will install a second, separate x86 installation of Homebrew, which is easiest done by running a shell entirely using Rosetta 2. Do so by running `arch -x86_64 zsh`. The remainder of the commands below should be run from within this `x86`-running window, though once the steps have been completed, the installed tools will work in any future shell.

3. Install a second installation of Homebrew for `x86` with `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`. It will automatically install itself into a second location (`/usr/local`, rather than `/opt/`).

4. Follow the same steps described in [Controlled Installation for macOS](https://leanprover-community.github.io/install/macos_details.html) using the `brew` you just installed:

```
/usr/local/bin/brew install elan mathlibtools
elan toolchain install stable 
elan default stable  
```

> 5. Install Visual Studio Code and the Lean extension via `brew install --cask visual-studio-code && code --install-extension jroesch.lean`.

Reviewer: Does it matter which `brew` you use here if you have both architectures installed?
```suggestion
5. Install Visual Studio Code and the Lean extension via `/usr/local/brew install --cask visual-studio-code && code --install-extension jroesch.lean`.
```
or:
```suggestion
5. Install Visual Studio Code and the Lean extension via `brew install --cask visual-studio-code && code --install-extension jroesch.lean` (both the x86 and ARM versions of `brew` will work).
```

Author: I imagine both should work - downloading from the website works too - but I haven't tested it.

ARM](https://github.com/actions/virtual-environments/issues/2187), installation
of Lean is for the moment a bit more complex.

Specifically, `elan` – which is installed
as part of the above instructions – will not be able to fetch Lean binaries on
these devices if installed the normal way.

The following instructions are adapted from [Fedor Pavutnitskiy](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/M1.20Macs.3A.20Installing.20the.20Lean.203.20toolchain/near/262832039), and allow you to install elan through [Rosetta](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment).

> 1. Open new terminal window and install xcode command line tools and rosetta 2 using ```xcode-select --install``` and ```softwareupdate --install-rosetta```.

Reviewer 2: ```suggestion
1. Open a new terminal window and install XCode Command Line Tools and Rosetta 2 using `xcode-select --install` and `softwareupdate --install-rosetta`.
```

> 2. Switch to the x86 architechture in the current terminal window by ```arch -x86_64 zsh ``` as boris suggests. Everything below should happen in the same window.

Reviewer 2: ```suggestion
2. We will install a second, separate x86 installation of Homebrew, which is easiest done by running a shell entirely using Rosetta 2. Do so by running `arch -x86_64 zsh`. The remainder of the commands below should be run from within this `x86`-running window, though once the steps have been completed, the installed tools will work in any future shell.
```

> 3. Install brew with ```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"```

Reviewer 2: ```suggestion
3. Install a second installation of Homebrew for `x86` with `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`. It will automatically install itself into a second location (`/usr/local`, rather than `/opt/`).
```

> 4. The same steps as described in [controlled installation for macos](https://leanprover-community.github.io/install/macos_details.html) using the brew you just installed:

Reviewer 2: ```suggestion
4. Follow the same steps described in [Controlled Installation for macOS](https://leanprover-community.github.io/install/macos_details.html) using the `brew` you just installed:
```

```
/usr/local/bin/brew install elan mathlibtools
elan toolchain install stable 
elan default stable  
```

> 5. Use ```leanproject get``` to get the repositories needed.

Reviewer 2: I'd remove this step and just point people at the Lean projects page at the end.

> 6. Install vs-code from the website and set it up as usual.

Reviewer 2: Here I think we should likely recommend installing VSCode via a `brew install --cask visual-studio-code && code --install-extension jroesch.lean` for the Lean extension.

Author: I installed it this way and it worked fine. Is there a reason this might be better, or work when the website download way does not?

Reviewer 2: Both should definitely work! The only reason I think this way is better is it's automated and doesn't involve clicking on things (i.e. someone can run it just the same way as the rest of the steps here).

Author: Good point, I guess as long as the user is on the command line, they can just install VSCode that way. I'll change this.

> ```leanproject``` should always be called from ```arch -x86_64```. Also, there may be some performance downsides to using lean under rosetta vs natively.

Reviewer 2: This shouldn't be true -- or if it is I think we should instead recommend installing `mathlibtools` with the ARM brew to avoid this, but yeah have you confirmed this is the case? What happens if you run it normally?

Author: Well I just wrote this because the person who wrote the instructions said it. Testing it out myself, I can run leanproject even without the arch -x86_64 and it runs pretty fast, faster than my non-M1 mac. I'll delete this line.
